src/filter.rs

```diff
@@ -1,168 +1,40 @@
 use std::fmt::{self, Display};
-use std::mem;
 use std::str::FromStr;
 
-use proc_macro2::Span;
-use syn::{File, Ident, Item, ItemConst, ItemFn, ItemMod, ItemType, TraitItem, Visibility};
+use syn::File;
 
 #[derive(Debug)]
 pub struct Filter {
-    segments: Vec<Ident>,
+    path: String,
 }
 
 impl FromStr for Filter {
-    type Err = String;
+    type Err = syn_select::Error;
 
     fn from_str(input: &str) -> Result<Self, Self::Err> {
-        let mut segments = Vec::new();
-        for segment in input.split("::") {
-            match syn::parse_str(segment) {
-                Ok(ident) => segments.push(ident),
-                Err(_) => return Err(format!("`{}` is not an identifier", segment)),
-            }
-        }
-        if segments.is_empty() {
-            return Err("empty path".to_owned());
-        }
-        Ok(Filter { segments })
+        let syntax_tree = File {
+            shebang: None,
+            attrs: Vec::new(),
+            items: Vec::new(),
+        };
+        syn_select::select(input, &syntax_tree)?;
```

Comment:
If you need to separate the filter from the search, I can make `syn_select::Selector` public.